### PR TITLE
Deliver requests asynchronously

### PR DESCRIPTION
**Pull request by acmcarther**
@alexpeachey

### CODE REVIEW

File: spec/replay_api/request_deliverer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      request_deliverer.connection = connection
+    end
+
+    describe '#post' do
+      let(:event) { Event.new }
+      let(:payload) { { replay_key: 'my_key' }.merge event.compact_attributes }
+
+      it 'sends the request' do
+        request_deliverer.post(uri:  '/events', payload: payload)
+        {} until request_deliverer.delivery_thread.status == "sleep"
```
Comment (acmcarther) [inline spec/replay_api/request_deliverer_spec.rb:21]: This is funny, but replacing `{}` with `sleep(x)` did not work.

File: lib/replay_api/request_deliverer.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    private
+
+    def start_delivery_thread
+      return if @delivery_thread
+
+      @delivery_thread = Thread.new do
+        loop do
+          requests = []
+          (requests << request_buffer.pop) until requests.length > 50 or time_since_last_write > 0.1
```
Comment (acmcarther) [inline lib/replay_api/request_deliverer.rb]: These magic values can be determined as being the request acceleration times some proportion of the request RTT. Need to play with this some more.

**Comment by acmcarther** [comment]
@alexpeachey Please review again. I intend to squash before merge as well.

File: lib/replay_api/client.rb [hunk 3]
```diff
@@ -6,37 +6,25 @@
...
+    def initialize(configuration=ReplayApi.configuration, deliverer=ReplayApi::RequestDeliverer.instance)
       @configuration ||= configuration
+      @deliverer ||= deliverer
     end
 
     def event(event=Event.new)
+      deliverer.post(uri: '/events', payload: payload(event.compact_attributes))
       yield event if block_given?
-      connection.post '/events', payload(event.compact_attributes)
     end
```
Comment (alexpeachey) [inline lib/replay_api/client.rb:19]: You changed this from posting after yield to posting before yield which means people using the block form can't fiddle with the attributes.

File: lib/replay_api/client.rb [hunk 4]
```diff
@@ -6,37 +6,25 @@
...
 
     def trait(trait=Trait.new)
+      deliverer.post(uri: '/traits', payload: payload(trait.compact_attributes))
       yield trait if block_given?
-      connection.post '/traits', payload(trait.compact_attributes)
     end
```
Comment (alexpeachey) [inline lib/replay_api/client.rb:24]: Same here.

File: lib/replay_api/client.rb (hunk 3 above)
Comment (acmcarther) [inline lib/replay_api/client.rb:19]: !! Good catch. I was fiddling around and saw this and misunderstood what it was doing. Quick fix.

File: spec/replay_api/client_spec.rb [hunk 5]
```diff
@@ -2,15 +2,12 @@
...
+    subject(:client) { Client.new(configuration, deliverer) }
+    let(:deliverer) { double :deliverer, post: nil }
     let(:configuration) { Configuration.new.tap { |c| c.replay_key = 'my_key' } }
-    before do
-      client.connection = connection
-    end
 
     describe '#event' do
-      let(:event) { Event.new }
+ deliverer     let(:event) { Event.new }
```
Comment (acmcarther) [inline spec/replay_api/client_spec.rb]: I see this, my bad.

File: lib/replay_api/request_deliverer.rb (hunk 2 above)
Comment (alexpeachey) [inline lib/replay_api/request_deliverer.rb]: Yeah, I would maybe make those numbers methods so it's clearer what they are and I would even move the logic to predicate methods.

```
def batch_full?(requests)
  requests.length > max_batch_size
end

def max_batch_size
  50
end

def max_delay_exeeded?
  (Time.now - time_last_write) > max_delay
end

def max_delay
  0.1
end
```

**Comment by bentona** [comment]
@alexpeachey how does this look to you? I'm going to merge this tomorrow if you guys don't have any more issues.

**Comment by troex** [comment]
Hmm have you tested it with rack server like unicorn, thin, puma? I see you run loop inside the thread but I don't see how you exit/stop the thread, won't that just hang the request inside the rack?

Maybe it won't block the request, but thread could keep running inside the rack server.

**Comment by bentona** [comment]
Good point @troex. Thoughts @acmcarther / @alexpeachey ?

**Comment by alexpeachey** [comment]
On Puma, I'm not sure what it will be like but probably much like Unicorn.
On Unicorn we need to make a bit in the readme about how they should initialize things up in the after_fork. Basically we don't want this running in the main Unicorn thread and we need one each running in each fork.

**Comment by acmcarther** [comment]
@bentona It looks like this doesn't have any issues running on a rack server ( as tested by @troex ). I'm not sure how to word the disclaimer in the readme. I'll clean up everything and then we can discuss what we want to say in the readme.

**Comment by bentona** [comment]
@alexpeachey Could you sketch out what a disclaimer like that would look like? I'm not quite sure how to convey that, just a bit more detail and we can write something up.

**Comment by alexpeachey** [comment]
Well you would need to build in the ability to disconnect or turn off the queue thingy. Then you make a command that could be called in the after_fork that turned it off in the main process and spun up a new one in the sub process. Then you would write docs about how that works and what to put in your unicorn.rb.

I know that's a lot of hand-waving but the answer needs research and trial/error to figure it out.

**Comment by troex** [comment]
I would suggest that maybe we don't need to initialize the deliverer until first event is send, that way queue/thread will not start in master process, and only start in child processes/forks. Than we don't need to care about `after_fork` or what rack server is used. @alexpeachey, @acmcarther your thoughts?

However I'm not sure about puma as it's different by design than other classic fork-based servers.

**Comment by acmcarther** [comment]
This is way out of my domain. I wasnt able to replicate any of these hypothetical issues on my test app. @troex can you throw up your test rack app? I can do some research, though if someone has more devops-y experience they might want to take the wheel.

**Pull request merged**